esphome/components/atm90e26/atm90e26.cpp

```diff
@@ -0,0 +1,234 @@
+
+  // Compute Checksum for the registers we set above
+  cs = ((volt_gain_ >> 8) + (volt_gain_ & 0xFF) + (ct_gain_ >> 8) + (ct_gain_ & 0xFF) + 0x75 + 0x30) & 0xFF;
+  cs |= ((volt_gain_ >> 8) ^ (volt_gain_ & 0xFF) ^ (ct_gain_ >> 8) ^ (ct_gain_ & 0xFF) ^ 0x75 ^ 0x30) << 8;
+  this->write16_(ATM90E26_REGISTER_CS2, cs);
+  ESP_LOGVV(TAG, "Set CS2 to: 0x%04X", cs);
+
+  this->write16_(ATM90E26_REGISTER_CALSTART,
+                 0x8765);  // Checks correctness of 21-2B registers and starts normal metering if ok
+  this->write16_(ATM90E26_REGISTER_ADJSTART,
+                 0x8765);  // Checks correctness of 31-3A registers and starts normal measurement  if ok
+
+  uint16_t sys_status = this->read16_(ATM90E26_REGISTER_SYSSTATUS);
+  if (sys_status & 0xC000) {  // Checksum 1 Error
+
+    ESP_LOGW(TAG, "Could not initialize ATM90E26 IC: CS1 was incorrect, expected: 0x%04X",
+             this->read16_(ATM90E26_REGISTER_CS1));
+    this->mark_failed();
+  }
+  if (sys_status & 0x3000) {  // Checksum 2 Error
+    ESP_LOGW(TAG, "Could not initialize ATM90E26 IC: CS2 was incorrect, expected: 0x%04X",
+             this->read16_(ATM90E26_REGISTER_CS2));
+    this->mark_failed();
+  }
+}
+
+void ATM90E26Component::dump_config() {
+  ESP_LOGCONFIG("", "ATM90E26:");
+  LOG_PIN("  CS Pin: ", this->cs_);
+  if (this->is_failed()) {
+    ESP_LOGE(TAG, "Communication with ATM90E26 failed!");
+  }
+  LOG_UPDATE_INTERVAL(this);
+  LOG_SENSOR("  ", "Voltage A", this->voltage_sensor_);
+  LOG_SENSOR("  ", "Current A", this->current_sensor_);
+  LOG_SENSOR("  ", "Power A", this->power_sensor_);
+  LOG_SENSOR("  ", "Reactive Power A", this->reactive_power_sensor_);
+  LOG_SENSOR("  ", "PF A", this->power_factor_sensor_);
+  LOG_SENSOR("  ", "Active Forward Energy A", this->forward_active_energy_sensor_);
+  LOG_SENSOR("  ", "Active Reverse Energy A", this->reverse_active_energy_sensor_);
+  LOG_SENSOR("  ", "Frequency", this->freq_sensor_);
+}
+float ATM90E26Component::get_setup_priority() const { return setup_priority::DATA; }
+
+uint16_t ATM90E26Component::read16_(uint8_t a_register) {
+  uint8_t data[2];
+  uint16_t output;
+
+  this->enable();
+  delayMicroseconds(4);
+  this->write_byte(a_register | 0x80);
+  delayMicroseconds(4);
+  this->read_array(data, 2);
+  this->disable();
+
+  output = (uint16_t(data[0] & 0xFF) << 8) | (data[1] & 0xFF);
+  ESP_LOGVV(TAG, "read16_ 0x%04X output 0x%04X", a_register, output);
+  return output;
+}
+
+void ATM90E26Component::write16_(uint8_t a_register, uint16_t val) {
+  ESP_LOGVV(TAG, "write16_ 0x%04X val 0x%04X", a_register, val);
+  this->enable();
+  delayMicroseconds(4);
+  this->write_byte(a_register & 0x7F);
+  delayMicroseconds(4);
+  this->write_byte((val >> 8) & 0xFF);
+  this->write_byte(val & 0xFF);
+  this->disable();
+}
+
+float ATM90E26Component::get_line_current_() {
+  uint16_t current = this->read16_(ATM90E26_REGISTER_IRMS);
+  return (float) current / 1000;
```

Comment:
Try changing these to divide by a float instead of casting them first.
```suggestion
  return current / 1000.0f;
```